chainstate/merkle_root.go

```diff
@@ -0,0 +1,77 @@
+package chainstate
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/BuxOrg/bux/utils"
+)
+
+type merkleRootProvider interface {
+	getName() string
+	verifyMerkleRoots(ctx context.Context, merkleProofs []string) error
+}
+
+// result of single merkle root verification by provider
+type verifyResult struct {
+	isError  bool
+	err      error
+	provider string
+}
+
+// VerifyMerkleRoots will try to verify merkle roots with all available providers
+func (c *Client) VerifyMerkleRoots(ctx context.Context, merkleRoots []string) error {
+	var wg sync.WaitGroup
+	resultsChannel := make(chan verifyResult)
+
+	for _, merkleRootPvdr := range createMerkleRootsProviders(c) {
+		wg.Add(1)
+		go func(provider merkleRootProvider) {
+			defer wg.Done()
+			verifyMerkleRootWithProvider(ctx, provider,
+				resultsChannel, merkleRoots)
+		}(merkleRootPvdr)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultsChannel)
+	}()
+
+	result := <-resultsChannel
+
+	if result.isError {
+		debugLog(c, "", fmt.Sprintf("verify merkle root error: %s from provider %s", result.err, result.provider))
```

Comment:
IMHO if the error is about connection to Pulse, we should rather log an Error here - because it's rather important that we can't connect Pulse.
And if the error is about not confirmed merkleroots, then we should log a Warning, because it is rather worth to store somewhere that someone tries to cheat on us.

Reply:
Done